### ISSUE TITLE
Map groups between Kirby and ResourceOwner

### ISSUE THREAD

**Issue by pReya** [thathoff/kirby-oauth#30]
Hey @thathoff,

thanks for your work and keeping this plugin updated <3.

I'd like to implement a use case, where SSO/OAuth users have different groups on the Identity Provider (Nextcloud via https://github.com/bahuma20/oauth2-nextcloud in my case).

Any chance you see a way to implement a simple mapping between groups? E.g. if `$provider->getResourceOwner($token)->getGroups()` ([documentation](https://github.com/bahuma20/oauth2-nextcloud/blob/master/src/Provider/NextcloudResourceOwner.php#L64) on the Provider side) contains `admin`, then assign Kirby group `admin`.

I'm not entirely sure how this could be configured in a nice and clean way. But this is quite a common use case for SSO/Oauth, right?

EDIT: For my use case a `groupWhitelist` could be enough, but it seems a proper mapping would still make sense as a general feature.

**Comment by thathoff** [comment]
You’re welcome! And it is a common use case indeed! I already thought about having a hook inside the `loginUser()` ([here](https://github.com/thathoff/kirby-oauth/blob/master/lib/Controller.php#L124)) method to implement custom modifications to the Kirby user based on the OAuth ResourceOwner.

**Comment by pReya** [comment]
Yeah, I guess a hook basically covers all use cases and is more flexible than hard coding specific user props. Any chance you can put the hook functionality on your roadmap?

**Comment by thathoff** [comment]
Since there is no roadmap for this plug-in I can't add it. But I think this makes the plug-in a lot more flexible this will be the next feature I'm going to implement. 

If that's not fast enough for your case or someone else likes to help here: Pull requests are very welcome... 😉

**Comment by pReya** [comment]
What would you suggest the function signature for this hook should look like? What are its parameters?

**Comment by pReya** [comment]
> I’m not sure if it should replace/bypass the whole [creation and check logic](https://github.com/thathoff/kirby-oauth/blob/master/lib/Controller.php?rgh-link-date=2023-02-16T09%3A00%3A19Z#L148) currently I would say it should be an addition.

This could also be another optional (boolean) parameter for the hook function, couldn't it?

**Comment by pReya** [comment]
@thathoff In the last year I've learned a lot more about Oauth, and I have an upcoming project, where I need to distinguish users based on their `groups` in the Identity Provider again.

I have thought about implementing this, but I'm currently a little lost regarding the PHP League Oauth Interface. The interface doesn't have any functions for getting custom claim (like `groups` from the token). So even with your suggested hook solution, we'd only have the `ResourceOwnerInterface` which doesn't offer anything to access custom claims from the token.

**Comment by thathoff** [comment]
This depends largely on your Oauth provider. I remember from a project that fetching the groups for a Google user is pretty hard to do. Have you tried to fetch the groups with the `$provider->getAuthenticatedRequest()` method and maybe an API providing the groups of the user? (see https://oauth2-client.thephpleague.com/usage/ for an example).

**Comment by thathoff** [comment]
I think the new [hooks release in 3.2.0](https://github.com/thathoff/kirby-oauth?tab=readme-ov-file#use-hooks) will solve this issue here, right? If not feel free to reopen the issue.
